### DISCUSSION TITLE
Combining the JUnit report.xml generated by running multiple GoogleTest executables

### DISCUSSION THREAD

**Discussion by prantoran** [google/googletest#4568] [Community Help]
Hello! 

I am new to GoogleTest and I have a number of cpp files each with its own GoogleTest initializations and `RUN_ALL_TESTS()` in the main function. I can generate their executables using Make build and individually generate report.xml for each of the test executables. 

However, I need to merge the generated report.xmls into a single report.xml that I can visualize in CICD pipeline.

Contents of the main function:
```cpp
int main(int argc, char **argv)
{
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
```

How I am generating a single report.xml:
```bash
./test_xxx --gtest_output="xml:report.xml"
```

**Comment by AeroSW** [comment]
## Comments on the problem

I am assuming that you are wanting to combine reports for reporting on **integration** test purposes.  If you are wanting to do unit tests for each system separately, but combine the results of the unit tests, then I would not recommend that.  I would recommend having unit tests write to their own reports to make it easier and faster for your pipeline to process results of each test individually.  You can use a separately running reporting app to accumulate the results from the individual tests for an overall report.  That way it can be executed as a standalone when you need to collect everything for a report.

## Description of my approach

I'm not sure if this is helpful to you, but if I am not mistaken, this is going to change based on how you open the file and write to it.  If you look through GTEST's source code, their reports are printed utilizing C's std::printf method.  I assume the `--gtest_output` is just redirecting the write buffer to be the file stream or fprintf.  Rather than utilizing GTEST's built in Environment Variable or its flag, you can write your own output that writes to a file.

I wrote my own, but this was due to me not being aware of them even having this flag.  My implementation follows the example in this [Stack Overflow Solution](https://stackoverflow.com/a/35186153/7044902).  ~~What you can do as well to prevent overriding the file's content is to offset the file write buffer to be at the current end of the file.~~ Looks  like for the code in this post, you should only need to add the `O_APPEND` flag to the usage of `open()` based on what is in `fcntl.h` documentation.

You are also able to take advantage of using `>>` in the cmd line where you are executing the tests to just append to the files. (Piping output to a file). [Append output to existing content of a file](https://stackoverflow.com/a/6065745/7044902)

I would have to dive into their usage of `--gtest_output` to know if there is an internal solution or not, but what I wrote above should be able to get you across a finish line relatively easily.

## Example from StackOverflow printf post
### My specs
| Spec | System |
| ---- | ----- |
| OS | Windows 11 x64 |
| Compiler | MinGW G++ 14.1.0 |
| STD Lib | C++23 |
### Example
Here is the code I am using in my unit tests currently. (Requires C++17 minimum, and I'm using C++23)
- [FileSystem Library (cppreference.com)](https://en.cppreference.com/w/cpp/filesystem)
- [fcntl.h (pubs.opengroup.org)](https://pubs.opengroup.org/onlinepubs/7908799/xsh/fcntl.h.html)
   - For appending, it looks like you would also want to use the `O_APPEND` flag in `open`.
```
#include <filesystem> // Used to ensure directory for output exists.
#include <fcntl.h>    // Contains libraries used from StackOverflow post.
namespace fs = std::filesystem;
void setup() {
    int fd;
    fs::path dir_path{"./output"};
    fs::path filename = dir_path / "tst-czengine.result.txt";
    if (!fs::is_directory(dir_path) && !fs::create_directory(dir_path)) {
        perror("Failed to find and create `./output`.");
        std::exit(1);
    }
    fd = open(filename.string().c_str(), O_WRONLY | O_CREAT, 0644);
    if (fd == -1) {
        perror("Open failed");
        std::exit(1);
    }
    if (dup2(fd, 1) == -1) {
        perror("dup2 failed");
        std::exit(1);
    }
}
```
